src/main/java/org/terasology/metalrenegades/interaction/systems/WellWaterSystem.java

```diff
@@ -15,36 +15,146 @@
+import org.terasology.dynamicCities.construction.events.BuildingEntitySpawnedEvent;
+import org.terasology.dynamicCities.settlements.SettlementEntityManager;
+import org.terasology.engine.Time;
 import org.terasology.entitySystem.entity.EntityManager;
 import org.terasology.entitySystem.entity.EntityRef;
 import org.terasology.entitySystem.event.ReceiveEvent;
 import org.terasology.entitySystem.systems.BaseComponentSystem;
 import org.terasology.entitySystem.systems.RegisterMode;
 import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.logic.characters.CharacterHeldItemComponent;
 import org.terasology.logic.common.ActivateEvent;
 import org.terasology.logic.inventory.events.GiveItemEvent;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.math.geom.Rect2i;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.metalrenegades.interaction.component.WaterCupComponent;
+import org.terasology.metalrenegades.interaction.component.WellBlockComponent;
 import org.terasology.metalrenegades.interaction.component.WellSourceComponent;
+import org.terasology.metalrenegades.interaction.events.CupFilledEvent;
+import org.terasology.metalrenegades.interaction.events.WellDrinkEvent;
+import org.terasology.metalrenegades.interaction.events.WellRefilledEvent;
 import org.terasology.registry.In;
+import org.terasology.thirst.component.ThirstComponent;
+import org.terasology.thirst.event.DrinkConsumedEvent;
+import org.terasology.world.time.WorldTimeEvent;
 
 /**
- * Tracks water source blocks inside of wells; and gives water cups to players upon interaction.
+ * Tracks water source blocks inside of wells, fills player's water cups upon interaction, and empties
+ * water cups upon consumption.
  */
 @RegisterSystem(RegisterMode.AUTHORITY)
 public class WellWaterSystem extends BaseComponentSystem {
 
+    private Logger logger = LoggerFactory.getLogger(WellWaterSystem.class);
+
+    private static final int CYCLES_UNTIL_REFILL = 20;
+    private int worldTimeCycles = 0;
+
     @In
     private EntityManager entityManager;
 
-    @ReceiveEvent(components = {WellSourceComponent.class})
-    public void onActivate(ActivateEvent event, EntityRef target) {
+    @In
+    private Time time;
+
+    @ReceiveEvent
+    public void onWorldTimeEvent(WorldTimeEvent worldTimeEvent, EntityRef entityRef) {
+        if (worldTimeCycles >= CYCLES_UNTIL_REFILL) {
+            for (EntityRef waterSource : entityManager.getEntitiesWith(WellSourceComponent.class)) {
+                WellSourceComponent wellSourceComp = waterSource.getComponent(WellSourceComponent.class);
+                wellSourceComp.waterRefills++;
+
+                if (wellSourceComp.waterRefills > wellSourceComp.maxRefills) {
+                    wellSourceComp.waterRefills = wellSourceComp.maxRefills;
+                }
+                waterSource.send(new WellRefilledEvent());
```

Comment:
this looks like you are sending the refilled event even though the well was already at maximum value 🧐 

```suggestion
                if (wellSourceComp.waterRefills < wellSourceComp.maxRefills) {
                    wellSourceComp.waterRefills++;
                    waterSource.send(new WellRefilledEvent());
                }
```

Reply:
Yep that is exactly what is happening here, nice catch!